strawberryfields/api/devicespec.py

```diff
@@ -15,62 +15,66 @@
 This module contains a class that represents the specifications of
 a device available via the API.
 """
-from collections.abc import Sequence
-import re
+from typing import Sequence, Mapping, Any
 
 import blackbird
 from blackbird.error import BlackbirdSyntaxError
 
 import strawberryfields as sf
 from strawberryfields.compilers import Ranges
-from strawberryfields.tdm.tdmprogram import TDMProgram
 
 
 class DeviceSpec:
     """The specifications for a specific hardware device.
 
     Args:
-        target (str): name of the target hardware device
         spec (dict): dictionary representing the raw device specification.
-            This dictionary should contain the following key-value pairs:
+            This dictionary must contain the following key-value pairs:
 
-            - layout (str): string containing the Blackbird circuit layout
+            - layout (str): string containing the Blackbird or XIR circuit layout
             - modes (int): number of modes supported by the target
             - compiler (list): list of supported compilers
             - gate_parameters (dict): parameters for the circuit gates
-
-        connection (strawberryfields.api.Connection): connection over which the
-            job is managed
     """
 
-    def __init__(self, target, spec, connection):
-        self._target = target
-        self._connection = connection
+    def __init__(self, spec: Mapping[str, Any]) -> None:
+        if not {"layout", "modes", "compiler", "gate_parameters"} <= spec.keys():
+            missing_keys = {
+                "target",
+                "layout",
+                "modes",
+                "compiler",
+                "gate_parameters",
+            } - spec.keys()
+            raise ValueError(
+                f"Device specification is missing the following keys: {sorted(missing_keys)}"
+            )
```

Comment:
I love the idea of the `if` condition here but it turns out to be excessive if we want to avoid duplication:
```suggestion
        missing_keys = {"target", "layout", "modes", "compiler", "gate_parameters"} - spec.keys()
        if missing_keys:
            raise ValueError(
                f"Device specification is missing the following keys: {sorted(missing_keys)}"
            )
```

Reply:
Completely right! I added the `missing_keys` later, so I missed this.